pkg/controller/route/routecontroller.go

```diff
@@ -26,86 +26,156 @@ import (
 	"k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/api/errors"
 	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/client/cache"
 	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
 	"k8s.io/kubernetes/pkg/cloudprovider"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/runtime"
 	"k8s.io/kubernetes/pkg/util/metrics"
 	nodeutil "k8s.io/kubernetes/pkg/util/node"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
 	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+	"k8s.io/kubernetes/pkg/watch"
 )
 
+type StatusType string
+
 const (
 	// Maximal number of concurrent CreateRoute API calls.
 	// TODO: This should be per-provider.
 	maxConcurrentRouteCreations int = 200
 	// Maximum number of retries of route creations.
 	maxRetries int = 5
 	// Maximum number of retries of node status update.
-	updateNodeStatusMaxRetries int = 3
+	updateNodeStatusMaxRetries int        = 3
+	statusDelete               StatusType = "Delete"
+	statusAdd                  StatusType = "Add"
+	statusUpdate               StatusType = "Update"
 )
 
 type RouteController struct {
 	routes      cloudprovider.Routes
 	kubeClient  clientset.Interface
 	clusterName string
 	clusterCIDR *net.IPNet
+	// Node framework and store
+	nodeController *framework.Controller
+	nodeStore      cache.StoreToNodeLister
+	// Nodes that need to be synced
+	queue *workqueue.Type
+	// To allow injection of syncRoutes for testing.
+	syncHandler func(nodeStatus NodeStatus) error
+	rateLimiter chan struct{}
+}
+
+type NodeStatus struct {
+	node   *api.Node
+	status StatusType
 }
 
-func New(routes cloudprovider.Routes, kubeClient clientset.Interface, clusterName string, clusterCIDR *net.IPNet) *RouteController {
+func New(routes cloudprovider.Routes, kubeClient clientset.Interface, clusterName string, clusterCIDR *net.IPNet, syncPeriod time.Duration) *RouteController {
 	if kubeClient != nil && kubeClient.Core().GetRESTClient().GetRateLimiter() != nil {
 		metrics.RegisterMetricAndTrackRateLimiterUsage("route_controller", kubeClient.Core().GetRESTClient().GetRateLimiter())
 	}
-	return &RouteController{
+	rc := &RouteController{
 		routes:      routes,
 		kubeClient:  kubeClient,
 		clusterName: clusterName,
 		clusterCIDR: clusterCIDR,
+		queue:       workqueue.New(),
+	}
+	rc.nodeStore.Store, rc.nodeController = framework.NewInformer(
```

Comment:
Instead of having a local cache, try to use a SharedInformer. The same cache will be used by at least the node and daemon set controllers.